GPU/Directx9/VertexShaderGeneratorDX9.cpp

```diff
@@ -715,7 +715,11 @@ void GenerateVertexShaderHLSL(const VShaderID &id, char *buffer, ShaderLanguage
 				break;
 
 			case GE_TEXMAP_ENVIRONMENT_MAP:  // Shade mapping - use dots from light sources.
-				WRITE(p, "  Out.v_texcoord = float3(u_uvscaleoffset.xy * float2(1.0 + dot(normalize(u_lightpos%i), worldnormal), 1.0 + dot(normalize(u_lightpos%i), worldnormal)) * 0.5, 1.0);\n", ls0, ls1);
+				{
+					std::string lightFactor0 = StringFromFormat("(length(u_lightpos%i) == 0.0 ? worldnormal.z : dot(normalize(u_lightpos%i), worldnormal))", ls0, ls0);
```

Comment:
Related to the previous comment, I wonder if GPU compilers optimize length==0 to avoid the square root. Then again, square roots are pretty fast on gpu so maybe it doesn't matter. No idea if either way is faster than just comparing all three components to 0.0.

Reply:
Hm, I can't get any even slightly consistent results between:

 * `all(equal(light.pos[%i], vec3(0.0)))`
 * `length(light.pos[%i]) == 0.0`
 * `dot(light.pos[%i], vec3(1.0)) == 0.0`
 * `dot(light.pos[%i], light.pos[%i]) == 0.0`
 * `light.pos[%i].x == 0.0 && light.pos[%i].y == 0.0 && light.pos[%i].z == 0.0`

If you have any preference, happy to change it.  No idea which might be faster on various drivers...

-[Unknown]

Reply:
Probably all of small enough impact to not really make a performance difference...